### ISSUE TITLE
Slow output with some Makefiles

### ISSUE THREAD

**Issue by danielhoherd** [sharkdp/bat#750]
The following Makefile pauses between each line of output. I have whittled it down to this string from a longer string, but could not whittle it down any further, so I don't know what part of the string is causing the problem.

```
FOO:=$($(BLARG_BUILD_MAKEFILE_PATH) $(BLARG_BUILD_MAKEFILE_PATH))
FOO:=$($(BLARG_BUILD_MAKEFILE_PATH) $(BLARG_BUILD_MAKEFILE_PATH))
FOO:=$($(BLARG_BUILD_MAKEFILE_PATH) $(BLARG_BUILD_MAKEFILE_PATH))
```

Also, the lines that take a long time to parse have a different colored `FOO` string than when you alter the line to where it does not have a delay problem.

```
$ time bat Makefile
───────┬──────────────────────────────────────────────────────────────────
       │ File: Makefile
───────┼──────────────────────────────────────────────────────────────────
   1   │ FOO:=$($(BLARG_BUILD_MAKEFILE_PATH) $(BLARG_BUILD_MAKEFILE_PATH))
   2   │ FOO:=$($(BLARG_BUILD_MAKEFILE_PATH) $(BLARG_BUILD_MAKEFILE_PATH))
   3   │ FOO:=$($(BLARG_BUILD_MAKEFILE_PATH) $(BLARG_BUILD_MAKEFILE_PATH))
───────┴──────────────────────────────────────────────────────────────────

real	0m2.834s
user	0m2.835s
sys	0m0.000s
```

This issue does not arise if you specify a different language, eg: `bat -l yaml Makefile`

This is similar to #677

**Comment by sharkdp** [comment]
thank you for reporting this. As with #677, this needs to be reported upstream.

**Comment by sharkdp** [comment]
I looked into this today. This seems to be caused by the following monster:

https://github.com/sublimehq/Packages/blob/f36b8f807d5f30d2b8ef639232a9fc5960f550fa/Makefile/Makefile.sublime-syntax#L40-L105

If we write out the `rule_lookahead` pattern, it looks like this:

```regex
[^()]*(?:([^()]*(?:([^()]*(?:([^()]*(?:([^()]*))?[^()]*))?[^()]*(?:([^()]*(?:([
^()]*))?[^()]*))?[^()]*))?[^()]*(?:([^()]*(?:([^()]*(?:([^()]*))?[^()]*))?[^()]
*(?:([^()]*(?:([^()]*))?[^()]*))?[^()]*(?:([^()]*(?:([^()]*))?[^()]*))?[^()]*))
?[^()]*))?[^()]*:(?!=)[^()]*(?:([^()]*(?:([^()]*(?:([^()]*(?:([^()]*))?[^()]*))
?[^()]*(?:([^()]*(?:([^()]*))?[^()]*))?[^()]*))?[^()]*(?:([^()]*(?:([^()]*(?:([
^()]*))?[^()]*))?[^()]*(?:([^()]*(?:([^()]*))?[^()]*))?[^()]*(?:([^()]*(?:([^()
]*))?[^()]*))?[^()]*))?[^()]*))?[^()]*
```

If we apply it to the line above
```
FOO:=$($(BLARG_BUILD_MAKEFILE_PATH) $(BLARG_BUILD_MAKEFILE_PATH))
```
there seems to be a form of catastrophic backtracking, as can be seen here: https://regex101.com/r/dTkE4u/1

**Comment by eth-p** [comment]
> ```regex
> [^()]*(?:([^()]*(?:([^()]*(?:([^()]*(?:([^()]*))?[^()]*))?[^()]*(?:([^()]*(?:([
> ^()]*))?[^()]*))?[^()]*))?[^()]*(?:([^()]*(?:([^()]*(?:([^()]*))?[^()]*))?[^()]
> *(?:([^()]*(?:([^()]*))?[^()]*))?[^()]*(?:([^()]*(?:([^()]*))?[^()]*))?[^()]*))
> ?[^()]*))?[^()]*:(?!=)[^()]*(?:([^()]*(?:([^()]*(?:([^()]*(?:([^()]*))?[^()]*))
> ?[^()]*(?:([^()]*(?:([^()]*))?[^()]*))?[^()]*))?[^()]*(?:([^()]*(?:([^()]*(?:([
> ^()]*))?[^()]*))?[^()]*(?:([^()]*(?:([^()]*))?[^()]*))?[^()]*(?:([^()]*(?:([^()
> ]*))?[^()]*))?[^()]*))?[^()]*))?[^()]*
> ```

Ouch. That looks like a backtracking nightmare. Do we know if it also catastrophically backtracks on Sublime?

**Comment by keith-hall** [comment]
@eth-p Sublime just doesn't use a backtracking regex engine (unless the regex pattern uses features that require backtracking, in which case it falls back to Onigurama) so this pattern can't cause catastrophic backtracking or performance issues in Sublime. I do think the C# regex pattern we fixed yesterday could be identified and optimized using static analysis, but I haven't yet looked into this Makefile one. It would be a lot of work though... At the very least, having an easy/automated way to spot patterns which could cause catastrophic backtracking could be a huge help with debugging (and when writing syntax definitions in the first place) and such a tool is on my todo list - no idea when I will get time to work on it though ;)

**Comment by sharkdp** [comment]
> unless the regex pattern uses features that require backtracking

what kind of features would require backtracking?

The regex posted above is actually just a part of other patterns. The full thing looks more like
```
(?=(?!(?x)[^()]*(?:\([^()]*(?:\([^()]*(?:\([^()]*(?:\([^()]*\))?[^()]*\))?[^()]
*(?:\([^()]*(?:\([^()]*\))?[^()]*\))?[^()]*\))?[^()]*(?:\([^()]*(?:\([^()]*(?:\
([^()]*\))?[^()]*\))?[^()]*(?:\([^()]*(?:\([^()]*\))?[^()]*\))?[^()]*(?:\([^()]
*(?:\([^()]*\))?[^()]*\))?[^()]*\))?[^()]*\))?[^()]*:(?!=)(?x)[^()]*(?:\([^()]*
(?:\([^()]*(?:\([^()]*(?:\([^()]*\))?[^()]*\))?[^()]*(?:\([^()]*(?:\([^()]*\))?
[^()]*\))?[^()]*\))?[^()]*(?:\([^()]*(?:\([^()]*(?:\([^()]*\))?[^()]*\))?[^()]*
(?:\([^()]*(?:\([^()]*\))?[^()]*\))?[^()]*(?:\([^()]*(?:\([^()]*\))?[^()]*\))?[
^()]*\))?[^()]*\))?[^()]*)(?x)[^()]*(?:\([^()]*(?:\([^()]*(?:\([^()]*(?:\([^()]
*\))?[^()]*\))?[^()]*(?:\([^()]*(?:\([^()]*\))?[^()]*\))?[^()]*\))?[^()]*(?:\([
^()]*(?:\([^()]*(?:\([^()]*\))?[^()]*\))?[^()]*(?:\([^()]*(?:\([^()]*\))?[^()]*
\))?[^()]*(?:\([^()]*(?:\([^()]*\))?[^()]*\))?[^()]*\))?[^()]*\))?[^()]*((\?|\+
|::?)?=|!=)(?x)[^()]*(?:\([^()]*(?:\([^()]*(?:\([^()]*(?:\([^()]*\))?[^()]*\))?
[^()]*(?:\([^()]*(?:\([^()]*\))?[^()]*\))?[^()]*\))?[^()]*(?:\([^()]*(?:\([^()]
*(?:\([^()]*\))?[^()]*\))?[^()]*(?:\([^()]*(?:\([^()]*\))?[^()]*\))?[^()]*(?:\(
[^()]*(?:\([^()]*\))?[^()]*\))?[^()]*\))?[^()]*\))?[^()]*|(?x)(?x)[^()]*(?:\([^
()]*(?:\([^()]*(?:\([^()]*(?:\([^()]*\))?[^()]*\))?[^()]*(?:\([^()]*(?:\([^()]*
\))?[^()]*\))?[^()]*\))?[^()]*(?:\([^()]*(?:\([^()]*(?:\([^()]*\))?[^()]*\))?[^
()]*(?:\([^()]*(?:\([^()]*\))?[^()]*\))?[^()]*(?:\([^()]*(?:\([^()]*\))?[^()]*\
))?[^()]*\))?[^()]*\))?[^()]*(\?|\+|::?)?=(?x)[^()]*(?:\([^()]*(?:\([^()]*(?:\(
[^()]*(?:\([^()]*\))?[^()]*\))?[^()]*(?:\([^()]*(?:\([^()]*\))?[^()]*\))?[^()]*
\))?[^()]*(?:\([^()]*(?:\([^()]*(?:\([^()]*\))?[^()]*\))?[^()]*(?:\([^()]*(?:\(
[^()]*\))?[^()]*\))?[^()]*(?:\([^()]*(?:\([^()]*\))?[^()]*\))?[^()]*\))?[^()]*\
))?[^()]*:(?!=)(?x)[^()]*(?:\([^()]*(?:\([^()]*(?:\([^()]*(?:\([^()]*\))?[^()]*
\))?[^()]*(?:\([^()]*(?:\([^()]*\))?[^()]*\))?[^()]*\))?[^()]*(?:\([^()]*(?:\([
^()]*(?:\([^()]*\))?[^()]*\))?[^()]*(?:\([^()]*(?:\([^()]*\))?[^()]*\))?[^()]*(
?:\([^()]*(?:\([^()]*\))?[^()]*\))?[^()]*\))?[^()]*\))?[^()]*)
```
and also includes positive and negative lookaheads.

**Comment by keith-hall** [comment]
Lookaheads are handled with the state-machine based regex engine in ST. Things like backreferences, recursion, atomic groups and lookbehinds currently make ST fall back to Onigurama. 
One can check within ST, if any patterns will cause Onigurama to be used - with a syntax definition open, choose the Syntax Tests build system and pick the `Regex Compatibility` variant.

**Comment by sharkdp** [comment]
I just tried to use the fancy-regex version of `bat`, and it is significantly faster (but still slow) on this test `Makefile` below:

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
onig | 4.054 ± 0.002 | 4.052 | 4.055 | 1.00 |
fancy-regex | 0.487 ± 0.000 | 0.487 | 0.487 | 1.00 |

``` Makefile
FOO:=$($(BLARG_BUILD_MAKEFILE_PATH) $(BLARG_BUILD_MAKEFILE_PATH))
FOO:=$($(BLARG_BUILD_MAKEFILE_PATH) $(BLARG_BUILD_MAKEFILE_PATH))
FOO:=$($(BLARG_BUILD_MAKEFILE_PATH) $(BLARG_BUILD_MAKEFILE_PATH))
FOO:=$($(BLARG_BUILD_MAKEFILE_PATH) $(BLARG_BUILD_MAKEFILE_PATH))
FOO:=$($(BLARG_BUILD_MAKEFILE_PATH) $(BLARG_BUILD_MAKEFILE_PATH))
```

**Comment by keith-hall** [comment]
You could try:

```yaml
  nps_unnested: '[^()]*'
  nps: '[^()]*(?=[()])'
  open: '(?:{{nps}}\('
  close: '\){{nps_unnested}})?'       # ignore this invalid.illegal
  just_eat: |         # WARNING: INSANITY FOLLOWS!
    (?x)(?:              # ignore whitespace in this regex
      {{open}}        # start level 1                      __
        {{open}}      # start level 2      ___/ _____ \__/ /
          {{open}}    # start level 3     is like snek... (by Valerie Haecky)
            {{open}}  # start level 4
              {{nps}} #       level 4
            {{close}} #   end level 4
          {{close}}   #   end level 3
          {{open}}    # start level 3
            {{open}}  # start level 4
              {{nps}} #       level 4
            {{close}} #   end level 4
          {{close}}   #   end level 3
          {{nps}}
        {{close}}     #   end level 2
        {{open}}      # start level 2
          {{open}}    # start level 3
            {{open}}  # start level 4
              {{nps}} #       level 4
            {{close}} #   end level 4
          {{nps}}
          {{close}}   #   end level 3
          {{open}}    # start level 3
            {{open}}  # start level 4
              {{nps}} #       level 4
            {{close}} #   end level 4
          {{nps}}
          {{close}}   #   end level 3
          {{open}}    # start level 3
            {{open}}  # start level 4
              {{nps}} #       level 4
            {{close}} #   end level 4
          {{nps}}
          {{close}}   #   end level 3
        {{nps}}
        {{close}}     #   end level 2
      {{nps}}
      {{close}}       #   end level 1
      |{{nps_unnested}})
```

I tried forcing Onigurama in Sublime (by adding `(?<=)` to the original monster pattern), and saw it was slow, then tweaked it with the above, and saw it was faster - it's still a monster though...

**Comment by sharkdp** [comment]
Using this replacement, highlighting that line is approximately 10,000 times faster :+1: :+1: 

Is the regex functionally equivalent? Do you think that this change has a chance of being merged upstream?

In any case, I'm going to create a patch for `bat` such that this issue can be closed.

Thank you very much, @keith-hall!

**Comment by keith-hall** [comment]
> Using this replacement, highlighting that line is approximately 10,000 times faster +1 +1
> Thank you very much

Yay, glad to hear it - you're welcome!

> Is the regex functionally equivalent? Do you think that this change has a chance of being merged upstream?

Yes, it is functionally equivalent as far as I can tell and reason about - plus it still passes all the syntax tests in Sublime. (Interestingly, forcing Oniguruma even on the original pattern causes 6 test assertion failures in Sublime - presumably it bails out once it takes a certain amount of time or backtracks a certain amount, I'm not sure.)

In my inexpert opinion, the Makefile syntax could be a good candidate for being re-written using the new non-deterministic parsing features available in Sublime. But, that said, I see no reason why this patch wouldn't be accepted, as it is functionally equivalent and even seems to increase performance slightly with ST's custom regex engine too. It may be an idea to try to tidy it up a bit though, to make the maintainer of the Makefile syntax happier :) (probably I removed some comments and messed up the pretty alignment a bit in my experimentation, and I ran out of time to beautify it)
